### PR TITLE
Return the oldval in monkeypatch for easier wrapping

### PR DESCRIPTION
**Pull request by HoverHell**

### CODE REVIEW

File: changelog/2505.trivial [hunk 1]
```diff
@@ -0,0 +1 @@
+Return the oldval in monkeypatch for easier wrapping.
```
Comment (The-Compiler) [inline changelog/2505.trivial]: This isn't very understandable I think - what about "Return the previous value in monkeypatch.setattr"?

Comment (HoverHell) [inline changelog/2505.trivial]: Changed that (for the changelog).

**Comment by HoverHell** [comment]
There could be a better way for monkeypatching a function with a wrapper of that function, but this is the smallest change useful for that.

**Comment by nicoddemus** [comment]
Thanks @HoverHell!

For completeness, I believe we should return `oldval` for all the other methods as well.

Also, this should target `features` given that is a new feature.

**Comment by RonnyPfannschmidt** [comment]
also it needs documentation, an example and tests

**Comment by HoverHell** [comment]
@nicoddemus 
@RonnyPfannschmidt

> For completeness

> documentation, an example and tests

I was hoping for a simple one-liner fix, but, well, since that does not seem to be accepted it's time to consider

> a better way for monkeypatching a function with a wrapper of that function

The main use case: I need to monkeypatch a method only partially, depending on the arguments.

Let's say I have a method `def _request(self, uri, ...)` (which could as well be a function / a classmethod / a staticmethod / a property (with values on `self` instead of arguments), etc.)

And let's say I want to wrap it like this:

    def _request_wrapping(self, uri, *args, **kwargs):
        if 'localhost' in uri:
            return _request_original(self, uri, *args, **kwargs)
        return FakeResponse()

Right now, as far as I see, I have to reproduce part of the logic of the `monkeypatch` regarding the various cases (property / classmethod / etc).

In a better case, `monkeypatch.setattr` would do the the wrapping.

The simple change in `monkeypatch.setattr` in this patch would make it possible to do

    def _request_wrapping(...):
        ...

    _request_original = monkeypatch.setattr(..., _request_wrapping)

A more general case would probably be thus:

   def _wrap(_request_original):

       def _request_wrapping(...):
           ...

       return _request_wrapping

   monkeypatch.setattr_f(..., func=_wrap)

where the `setattr_f` would do `setattr(target, name, func(oldval))`.

And, for usage simplicity (to avoid double nesting), a wrapper-maker like

    def make_kwarg_wrapper(wrapping, value_kwarg="_wrapped_value"):

        @functools.wraps(wrapping)
        def wrapper(func_to_wrap):

            @functools.wraps(func_to_wrap)
            def outer_wrapping(*args, **kwargs):
                kwargs[value_kwarg] = func_to_wrap
                return wrapping(*args, **kwargs)

            return outer_wrapping

        return wrapper

Question is: which way seems to be better, and which ways am I not noticing?

**Comment by nicoddemus** [comment]
> The main use case: I need to monkeypatch a method only partially, depending on the arguments.

I can't think atm how to support your use case directly in a clean manner. Your case depends on the arguments passed to the patched function to decide to either call the original function or return some mock object, and all the APIs mentioned above and the ones I can think of feel clumsy to me.

FWIW I usually switch over to `mock` (with `pytest-mock`) when I need more complex patching. I'm not sure we want to support more ways to patch things given that `mock` already has quite a huge API which supports many use cases and is part of the standard library.

Perhaps we should just follow with your original approach for now? The implementation is really simple and updating the docstrings to mention that now the functions return the patched object and a few regression tests also seem quick to accomplish.

Then your use case can be implemented as:

```python
def request_wrapping(self, uri, *args, **kwargs):
    if 'localhost' in uri:
        return request_original(self, uri, *args, **kwargs)
    return FakeResponse()

request_original = monkeypatch.setattr(module, 'request', request_wrapping)
```

Which is not actually that bad IMHO, given the complexity of the patching in this case.

**Comment by RonnyPfannschmidt** [comment]
hmm, im pondering whether this is a good idea to begin with, 

an monkeypatch.decorate/decorate_item may be a good helper isntead,

then the writing would be

```
@monkeypatch.decorate('foo.bar')
def _wrap_fun(f):
  @functools.wraps(f)
  def new(...):
     return f(...)
```

**Comment by HoverHell** [comment]
> an monkeypatch.decorate/decorate_item may be a good helper isntead,

That's the tricky question: since a one-liner fix is not okay, it's better to think what more general solution is better (even if it plugs into the same modification inside).

For now, I'm using a partial (and potentially buggy) copypaste of that method:

    def _wrap(target, name, func, _orig_kwarg="_sup", mp=monkeypatch):
        oldval = target.__dict__.get(name, None)  # incomplete piece of logic from `mp.setattr`

        @functools.wraps(func)
        def wrapped(*args, **kwargs):
            kwargs[_orig_kwarg] = oldval
            return func(*args, **kwargs)

        mp.setattr(target, name, wrapped)
        return wrapped

the `decorate` way is a bit too bulky, which becomes a problem when there is a need to override 10+ methods (for upper-level overall-testing of glue-code)

**Comment by nicoddemus** [comment]
> That's the tricky question: since a one-liner fix is not okay, 

I only think we need to return `oldval` for all variants, implement tests and update the docs. That's a requirement for all code patches IMHO.

> it's better to think what more general solution is better (even if it plugs into the same modification inside).

I'm not sure guys, I feel we would be duplicating advanced functionality which is already available in the builtin `mock` module. As said [in the comment above](https://github.com/pytest-dev/pytest/pull/2505#issuecomment-309931571), it is not so easy to implement this given that the requirements can vary a lot from case to case.

Btw, is there something you guys don't like about the example showing how to achieve the same using `mock`, or it is because you feel `monkeypatch` should support this out of the box?

**Comment by RonnyPfannschmidt** [comment]
from my pov having to use mock is a bug, the only valid case is if you cant yet abandon a legacy api

**Comment by nicoddemus** [comment]
@HoverHell gentle ping, are you still interested in this?

**Comment by RonnyPfannschmidt** [comment]
closing for now, please let us know if you want t continue

**Pull request closed without merging**